lib/EnsEMBL/REST/Controller/Sequence.pm

```diff
@@ -265,6 +308,119 @@ sub _process_feature {
+sub _check_limits {
+  my ($self, $c, $seq_len) = @_;
+
+  # Oh the fun of zero indexing strings in a one-indexed sub-sequence world
+  my $start = $c->request->param('start') ? $c->request->param('start') : 1;
+  my $end = defined($c->request->param('end')) ? $c->request->param('end') : $seq_len;
+
+  # Sanity checking
+  if($start < 0 || $start > $seq_len) {
+      Catalyst::Exception->throw(qq{Your start coordinate is not within the sequence requested})
+  }
+  if($end < 1 || $end > $seq_len + 1) {
+      Catalyst::Exception->throw(qq{Your end coordinate is not within the sequence requested})
+  }
+  if($start > $end) {
+      Catalyst::Exception->throw(qq{Your start coordinate cannot be larger than your end})
+  }
+
+  return ($start, $end);
+}
+
+# For grabbing the sub-sequence for a protein sequence from
+# a translation object. The object type in our stash must
+# be a Bio::EnsEMBL::PredictionTranscript or Bio::EnsEMBL::Transcript
+sub _translate_coordinates {
+  my ($self, $c, $obj) = @_;
+
+  # Return if we've already translated the coordinates
+  return if($c->{stash}->{coordstranslated});
+
+  my $seq_len = length $obj->spliced_seq();
+  my $start; my $end;
+  if($obj->strand() == 1) {
+      $start = $c->request->param('start') ? $c->request->param('start') + $obj->seq_region_start() - 1 : $obj->seq_region_start();
+      $end = $c->request->param('end') ? $obj->seq_region_start() + $c->request->param('end') : $obj->seq_region_end();
+  } else {
+      # Things get a little messy if we're on the reverse strand, we have to count from
+      # opposite ends.
+      $end = $c->request->param('start') ? $obj->seq_region_end() - $c->request->param('start') : $obj->seq_region_end();
+      $start = $c->request->param('end') ? $obj->seq_region_start() + $seq_len - $c->request->param('end') : $obj->seq_region_start();
+  }
+
+  # Do we have a translation?
+  return unless($obj->translate());
+
+  my $transcript_mapper = $obj->get_TranscriptMapper();
+
+  # Grab the coordinates for the peptide sequence that maps from
+  # this subsequence of the transcript
+  my @coords = $transcript_mapper->genomic2pep($start, $end, $obj->strand());
+
+  # Go through and build the coordinates from the pieces returned
+  my $pep_start = length $obj->translate()->seq(); my $pep_end = 0;
+  foreach my $coord (@coords) {
+      if($coord->isa('Bio::EnsEMBL::Mapper::Coordinate')) {
+	  $pep_start = $coord->start if($coord->start < $pep_start);
+	  $pep_end = $coord->end if($coord->end > $pep_end);
+      }
+  }
+
+  # See if we don't find any peptide sequence in this window, so
+  # no coordinates should be returned.
+  if($pep_end < $pep_start) {
+      $c->request->params->{'start'} = 0;
+      $c->request->params->{'end'} = 0;
+  } else {
+      $c->request->params->{'start'} = $pep_start;
+      $c->request->params->{'end'} = $pep_end;
+  }
+
+  # Remember we've translated the coordinates so we don't try again
+  $c->{stash}->{coordstranslated} = 1;
```

Comment:
Direct access of stash hash is a bit naughty. These should be set via the accessor. $c->stash->{coordstranslated} = 1;